### PR TITLE
[SPARK-48392][CORE] Also load `spark-defaults.conf` when provided `--properties-file`

### PR DESCRIPTION
**Pull request by sunchao**
### What changes were proposed in this pull request?

Currently if a property file is provided as argument to Spark submit, the `spark-defaults.conf` is ignored. This PR changes the behavior such that `spark-defaults.conf` is still loaded in this scenario, and any Spark configurations that are in the file but not in the input property file will be loaded.

### Why are the changes needed?

Currently if a property file is provided as argument to Spark submit, the `spark-defaults.conf` is ignored. This causes inconveniences for users who want to split Spark configurations into the two. For example, in Spark on K8S users may want to store system wide default settings in `spark-defaults.conf`, while user specified configurations that are more dynamic in an property file and pass to Spark applications via `--properties-file` parameter. Currently this is not possible. See also https://github.com/kubeflow/spark-operator/issues/1321.

### Does this PR introduce _any_ user-facing change?

Yes, now when a property file is provided via `--properties-file`, `spark-defaults.conf` will also be loaded. However, those configurations specified in the former will take precedence over the same configurations in the latter.

### How was this patch tested?

Existing tests and a new test.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by sunchao** [comment]
cc @viirya @dongjoon-hyun @cloud-fan let me know what you think

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala [hunk 1]
```diff
@@ -125,14 +125,27 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
...
-    // Use common defaults file, if not specified by user
-    propertiesFile = Option(propertiesFile).getOrElse(Utils.getDefaultPropertiesFile(env))
     // Honor --conf before the defaults file
     defaultSparkProperties.foreach { case (k, v) =>
       if (!sparkProperties.contains(k)) {
         sparkProperties(k) = v
       }
     }
+
+    // Also load properties from `spark-defaults.conf` if they do not exist in the properties file
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala:135]:
```suggestion
    // Also load properties from `spark-defaults.conf` if they do not exist in the properties file
    // and --conf list
```

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala [hunk 2]
```diff
@@ -125,14 +125,27 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
...
+      val properties = Utils.getPropertiesFromFile(filename)
+      properties.foreach { case (k, v) =>
+        if (!sparkProperties.contains(k)) {
+          sparkProperties(k) = v
+        }
+      }
+    }
+
+    if (propertiesFile == null) {
+      propertiesFile = defaultSparkConf
```
[the remark this answers is not in the document]
Comment (viirya) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala:147]: I think this follows original behavior to overwrite `propertiesFile` value. Although it looks like a strange behavior.

**Review by viirya — COMMENTED** [review]
I think this proposed behavior looks reasonable to me. If there is any document on this, we probably need to update it too.

Do you also think we should add an item to migration guide for this?

**Comment by sunchao** [comment]
> I think this proposed behavior looks reasonable to me. If there is any document on this, we probably need to update it too.

Thanks @viirya . There is a related documentation here: https://spark.apache.org/docs/latest/configuration.html#dynamically-loading-spark-properties. It looks like the behavior change in this PR is consistent with the description there:

> Any values specified as flags or in the properties file will be passed on to the application and merged with those specified through SparkConf. Properties set directly on the SparkConf take highest precedence, then flags passed to spark-submit or spark-shell, then options in the spark-defaults.conf file.

> Do you also think we should add an item to migration guide for this?

Yes sure. I can add this to the migration guide as a separate PR afterwards.

**Comment by viirya** [comment]
> Thanks @viirya . There is a related documentation here: https://spark.apache.org/docs/latest/configuration.html#dynamically-loading-spark-properties. It looks like the behavior change in this PR is consistent with the description there:
> 

Yea, but seems it doesn't mention how does `--properties-file` work with `spark-defaults.conf`. Maybe we can add few words there too.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala [hunk 3]
```diff
@@ -125,14 +125,27 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
    * When this is called, `sparkProperties` is already filled with configs from the latter.
    */
   private def mergeDefaultSparkProperties(): Unit = {
-    // Use common defaults file, if not specified by user
-    propertiesFile = Option(propertiesFile).getOrElse(Utils.getDefaultPropertiesFile(env))
     // Honor --conf before the defaults file
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala:128]:
```suggestion
    // Honor --conf before the specified properties file and defaults file
```

**Review by viirya — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala (hunk 3 above)
Comment (sunchao) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: Done

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala (hunk 1 above)
Comment (sunchao) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala:135]: Done

**Comment by sunchao** [comment]
Yea makes sense. Just updated the doc too.

**Comment by sunchao** [comment]
Thanks! merged to master/4.0

**Comment by yaooqinn** [comment]
You need to update all the usage information for commands with `--help`

**Comment by sunchao** [comment]
@yaooqinn do you mean the `--help` from `spark-submit` / `spark-shell`? currently the usage about `--properties-file` is as follow:

```
  --properties-file FILE      Path to a file from which to load extra properties. If not
                              specified, this will look for conf/spark-defaults.conf.
```

which I think is also consistent with the behavior here although not complete? I can add more info there.

**Comment by yaooqinn** [comment]
```
--properties-file FILE   Path to a custom Spark properties file. Default is conf/spark-defaults.conf.
```

Do the commands with the above message also get affected?

**Comment by sunchao** [comment]
No they are not affected. The change in this PR is only in `SparkSubmitArguments`. Classes such as `HistoryServerArguments` calls `Utils.loadDefaultSparkProperties`.

**Comment by yaooqinn** [comment]
I'm not sure if such an implicit discrepancy is user-friendly or not.

**Review by advancedxy — COMMENTED** [review]
This is a behavior change and it seems there's no easy way for users to restore the original behavior. Do you think it would be safer to add a new cli argument, such as `--extra-properties-file` instead? By adding `--extra-properties-files`, I think the affect scope could be minimum. 

A potential use case for the original `--properties-file`: there might be multiple spark clusters(or yarn clusters) to submit spark jobs and there are multiple spark-config files under $SPARK_CONF_DIR, such as:
1. spark-defaults.conf --> which stores the default cluster settings
2. spark-cluster-a.conf --> which stores the cluster settings for cluster a
3. spark-cluster-b.conf --> which stores the cluster settings for cluster b

For most cases, the default cluster is used. For some specific jobs, cluster-a might be used. After this change, both `spark-defaults.conf` and `spark-cluster-a.conf` are loaded, which might be undesired as users are already assuming only one property file will be loaded.

**Comment by viirya** [comment]
> * spark-defaults.conf --> which stores the default cluster settings
> * spark-cluster-a.conf --> which stores the cluster settings for cluster a
> * spark-cluster-b.conf --> which stores the cluster settings for cluster b

For the properties defined in `spark-cluster-a.conf`, it will be effective. So any duplicate properties defined in `spark-defaults.conf` as default ones, will be overwritten. I think this still follows original behavior.

For what is described, it seems giving this change more support? As users can put common default ones in `spark-defaults.conf` and apply them to all clusters. Only the different properties need to be defined in separate files.

**Comment by sunchao** [comment]
+1 on what @viirya said. 

@advancedxy to be clear the configurations from `spark-cluster-a.conf` will still take higher precedence than those from `spark-defaults.conf`, and I do think this is more aligned with the purpose of `spark-defaults.conf`. This also addresses an inconsistency in the current implementation: for configurations passed through `--conf` flags, Spark will still load `spark-defaults.conf`, but it will not for those passed from `--properties-file`, which is a bit confusing.  

Since we are preparing a new major release, I'm just hoping to take the opportunity to improve this. This is also important for kubeflow [Spark k8s operator](https://github.com/kubeflow/spark-operator) users, since currently there is no way to separate system-wide default configurations and user and job specific configurations.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala [hunk 4]
```diff
@@ -125,14 +125,28 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
    * When this is called, `sparkProperties` is already filled with configs from the latter.
    */
   private def mergeDefaultSparkProperties(): Unit = {
-    // Use common defaults file, if not specified by user
-    propertiesFile = Option(propertiesFile).getOrElse(Utils.getDefaultPropertiesFile(env))
-    // Honor --conf before the defaults file
+    // Honor --conf before the specified properties file and defaults file
     defaultSparkProperties.foreach { case (k, v) =>
```
[the remark this answers is not in the document]
Comment (sunchao) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala:129]: Oops that's right. I'll try to address it in the follow-up PR.

**Comment by cloud-fan** [comment]
IIUC, we want to have a way to load the default config files, and it's cleaner if `--properties-file` can do it. However, I do share the same concern of @advancedxy . Adding `--extra-properties-files` won't hurt and you can use this new option to load the default config files. Then we can leave `--properties-file` unchanged and avoid the risk of breaking Spark users during version upgrade.

**Comment by cloud-fan** [comment]
BTW, I agree that it's a better behavior to just let `--properties-file` load the default config file. We should have used this behavior on the first day. But now I think it's a bit overkill to make a breaking change just for a cleaner API. `--extra-properties-files` should be sufficient functional wise.

**Comment by yaooqinn** [comment]
+1 for `--extra-properties-files`.

As a maintainer of Apache Kyuubi, which supports multitenant and multi-runtime environments. This change might pollute an isolated custom spark conf.

**Comment by sunchao** [comment]
Thanks for all the feedback! Let me re-purpose the follow up PR #46782 to add this extra flag then (and we no longer need to update the migration guide). One minor inconvenience is we'd also need to update kubeflow k8s operator to consume this flag, and then users of the operator need to upgrade first in order to use this. I'll make a PR there too.

**Pull request closed without merging**